Prot Warrior Beta Build 45969 Changes/Fixes

Wyste opened the pull request:
Build 45969 changes:
- Avatar updated to gain 10 rage instead of 15, and handle just the tier28 outburst mechanic
- Booming Voice talent gives 30 rage now instead of 40
- Removed everything related to Wrenching Impact being removed from game/build.
- Changed Heroic Throw cooldown to 1sec
- Thunder Clap and Shield Slam modified to give 50% more rage for violent_outburst being up (was flat 10/30 rage respectively)
- Thunder Clap and Shield Slam modified to handle tier28 and talent outburst buffs separately
- Titanic Throw cooldown updated to 3 seconds
- Added hasteCD to Thunder Clap
- Extend Ravager's duration by 2 sec (hasted) using Dance of Death talent
- Impenetrable Wall – Now grants Shield Slam 3 extra Rage (was 5).
- Heavy Repercussions – Now grants Shield Slam 2 extra Rage (was 3).
- Champion’s Bulwark – Grants 20 bonus Rage (was 30).
- Changed Thunderlord's max CD reduction to 3 seconds from 4

Fixes:
- Rename tier28 outburst to outburst and outburst_buff due to the DF talent being called violent_outburst now
- UNIT_POWER_FREQUENT updated to handle either tier28 or violent outburst talent
- Hook "spend"  updated to handle either tier28 or violent outburst talent

Additional Fix:
- Fix for unnerving focus not calculating properly and throwing error

Wyste committed:
Build 45969 changes + additional fixes

Build 45969 changes:
- avatar updated to gain 10 rage instead of 15, and handle JUST the tier28 outburst mechanic
- booming voice talent gives 30 rage now instead of 40
- removed everything related to wrenching impact being removed
- changed heroic throw cooldown to 1sec
- shield charge gives 20 rage instead of 30
- thunder clap and shield slam modified to give 50% more rage for violent_outburst being up (was flat 10/30 rage respectively)
- thunder clap and shield slam modified to handle tier28 and talent outburst buffs separately
- titanic_throw cooldown updated to 3 seconds
- added hasteCD to tclap
- extend ravager's duration by 2 sec (hasted) using dance of death talent
- Impenetrable Wall – Now grants Shield Slam 3 extra Rage (was 5).
- Heavy Repercussions – Now grants Shield Slam 2 extra Rage (was 3).
- Champion’s Bulwark – Grants 20 bonus Rage (was 30).
- changed thunderlord's max CD reduction to 3 seconds from 4

Fixes:
- rename tier28 outburst to outburst and outburst_buff due to the DF talent being called violent_outburst now)
- UNIT_POWER_FREQUENT updated to handle either tier28 or violent outburst talent enabled
- hook "spend"  updated to handle either tier28 or violent outburst talent enabled

Wyste committed:
Fix unnerving focus rage multiplier

Hekili commented on Dragonflight/WarriorProtection.lua:
```diff
@@ -349,7 +348,7 @@ spec:RegisterAuras( {
     ravager = {
         id = 228920,
-        duration = 12,
+        duration = function () return (12 + (buff.dance_of_death.up and 2 or 0)) * haste end,
```
`return ( buff.dance_of_death.up and 14 or 12 ) * haste`

Wyste replied:
Fixed in https://github.com/Wyste/hekili/commit/101bcd4e90bfb6ffe25dd8e3fb43c7209307c34e

Hekili commented on Dragonflight/WarriorProtection.lua:
```diff
@@ -517,8 +516,8 @@ spec:RegisterUnitEvent( "UNIT_POWER_FREQUENT", "player", nil, function( event, u
 
-            if state.set_bonus.tier28_2pc > 0 or state.talent.outburst.enabled then
-                outburstRage = ( outburstRage + lastRage - current ) % 30 -- Outburst.
+            if state.set_bonus.tier28_2pc > 0 or state.talent.violent_outburst.enabled then
```
Risk:  If there are mechanics that reset this rage value (i.e., reset on boss pull, reset on death), that needs to be addressed somewhere.  (There's a similar issue with Marksmanship's Sanctum tier set.)

Wyste replied:
Verified in-game with several situations:
Fresh Login, fresh spec talent refresh (application), and reequipped all 4 tier pieces.
- Combat engaged in Maldraxxus, Used a single revenge (20/30 rage;  no 1st seeing red stack yet).
- Killed mob, Dala hearthstoned, used Orgrimmar portal, and flew into Durotar, engaged mobs using only shield slam until enough rage, then Revenged again, gained 1 stack of Seeing Red (expected, as I'd be 40/30 at this point).
- Flew into Org, engaged Training dummy, used Revenge and procced another stack of seeing red (10 excess from durotar, + 20 from new revenge). 
- Ran same test 3 other times through various zones and through load screens (across diff content) with same results.

From these kinds of tests it is my assumption that rage spent is tracked through zones/load screens as well, no resets that I'm seeing anywhere.

Note: I did not have time to test if the outburst rage needs reset on a boss pull, and I won't have an opportunity to see what tier28 does in retail until next week :(

Will fix some minor syntax stuff related to outburst in upcoming commit.

Wyste replied:
Fixed in https://github.com/Wyste/hekili/commit/101bcd4e90bfb6ffe25dd8e3fb43c7209307c34e

Hekili commented on Dragonflight/WarriorProtection.lua:
```diff
@@ -891,7 +890,7 @@ spec:RegisterAbilities( {
         id = 6544,
         cast = 0,
-        cooldown = function () return 45 + (talent.bounding_stride.enabled and -15 or 0) + (talent.wrenching_impact.enabled and 45 or 0) end,
+        cooldown = function () return 45 + (talent.bounding_stride.enabled and -15 or 0) end,
```
Formatting for braces in addon code is to space them out from a consistency standpoint.
i.e.  `( talent.bounding_stride.enabled and -15 or 0 )`
In this case, I would do:  `return talent.bounding_stride.enabled and 30 or 45`

Wyste replied:
Fixed in https://github.com/Wyste/hekili/commit/101bcd4e90bfb6ffe25dd8e3fb43c7209307c34e

Hekili commented on Dragonflight/WarriorProtection.lua:
```diff
@@ -1264,7 +1263,12 @@ spec:RegisterAbilities( {
         gcd = "spell",
 
-        spend = function () return ( 15 + ( talent.impenetrable_wall.enabled and -5 or 0 ) + ( talent.heavy_repercussions.enabled and -18 or -15 ) ) * ( talent.unnerving_focus.enabled and 1.5 or 1) end,
+        spend = function () return 
```
It's important to validate whether multipliers impact only the base value or the modifications from talents as well.  Have you tested the actual in-game combinations of these talents/buffs to verify?

Is it correct that Violent Outburst and Unnerving Focus increase the cost?

Wyste replied:
I have verified that outburst enhances the rage gain on both Tclap and Shield Slam based on both outburst, unnerving focus, and both auras being up at the same time. [WCL Link](https://www.warcraftlogs.com/reports/kgNd3pFLnMwYHfrW/#type=resources&spell=101&source=7&view=events&boss=0&difficulty=0) . I am having difficulty tracking down the math definition in wow.tools or wowhead , so I'm guessing it's part of the server script portion.

Wyste replied:
Fixed in https://github.com/Wyste/hekili/commit/101bcd4e90bfb6ffe25dd8e3fb43c7209307c34e

Hekili commented on Dragonflight/WarriorProtection.lua:
```diff
@@ -1251,7 +1250,7 @@ spec:RegisterAbilities( {
             if talent.champions_bulwarkj.enabled then
```
Typo:  `champions_bulwark`

Hekili commented on Dragonflight/WarriorProtection.lua:
```diff
@@ -1463,19 +1471,19 @@ spec:RegisterAbilities( {
             end
 
-            if buff.outburst.up then
+            if buff.violent_outburst.up then
```
Is overlap between T28 and talent tested?  If you have Violent Outburst and Outburst, are both consumed, or just one of them?  As written, only one of the buffs is consumed here.  Is that correct?

Wyste replied:
Looks like T28 and Talent are BOTH consumed. Tested in-game as well and can verify that both buffs are consumed (if they are up).

![image](https://user-images.githubusercontent.com/5103286/194471791-941058d4-70f0-4d57-b088-c9dd6523336a.png)

Wyste replied:
Fixed in https://github.com/Wyste/hekili/commit/101bcd4e90bfb6ffe25dd8e3fb43c7209307c34e

Wyste committed:
Hekili comment review + fixes

Added Unnerving Focus to Demoralizing Shout (Booming Voice) rage gain.
Modified syntax of rage spenders / gains to conform to Hekili's syntax.
Syntax correction for ravager's duration made
Outburst rage tracker logic modified.
Typo's fixed.
Fixed Shield Slams spender to be negetive (gain rage) and handle both forms of outburst as independant if blocks.

Addresses https://github.com/Hekili/hekili/pull/1642#discussion_r989204340
Addresses https://github.com/Hekili/hekili/pull/1642#discussion_r989202400
Addresses https://github.com/Hekili/hekili/pull/1642#discussion_r989202126
Addresses https://github.com/Hekili/hekili/pull/1642#discussion_r989198190
Addresses https://github.com/Hekili/hekili/pull/1642#discussion_r989196182
Addresses https://github.com/Hekili/hekili/pull/1642#discussion_r989194838

Wyste commented:
@Hekili  Please re-review. Thanks again.
